File: src/bun.js/test/expect/toHaveBeenLastCalledWith.zig
pub fn toHaveBeenLastCalledWith(this: *Expect, globalThis: *JSGlobalObject, callframe: *CallFrame) bun.JSError!JSValue {
    jsc.markBinding(@src());

    const thisValue = callframe.this();
    const arguments = callframe.arguments();
    defer this.postMatch(globalThis);
    const value: JSValue = try this.getValue(globalThis, thisValue, "toHaveBeenLastCalledWith", "<green>...expected<r>");

    this.incrementExpectCallCounter();

    const calls = try bun.cpp.JSMockFunction__getCalls(globalThis, value);
    if (!calls.jsType().isArray()) {
        return this.throw(globalThis, comptime getSignature("toHaveBeenLastCalledWith", "<green>...expected<r>", false), "\n\nMatcher error: <red>received<r> value must be a mock function\nReceived: {any}", .{value.toJestPrettyFormat(globalThis)});
    }

    const totalCalls: u32 = @truncate(try calls.getLength(globalThis));
    var lastCallValue: JSValue = .zero;

    var pass = totalCalls > 0;

    if (pass) {
        lastCallValue = try calls.getIndex(globalThis, totalCalls - 1);

        if (!lastCallValue.jsType().isArray()) {
            return globalThis.throw("Expected value must be a mock function with calls: {any}", .{value.toJestPrettyFormat(globalThis)});
        }

        if (try lastCallValue.getLength(globalThis) != arguments.len) {
            pass = false;
        } else {
            var itr = try lastCallValue.arrayIterator(globalThis);
            while (try itr.next()) |callArg| {
                if (!try callArg.jestDeepEquals(arguments[itr.i - 1], globalThis)) {
                    pass = false;
                    break;
                }
            }
        }
    }

    if (pass != this.flags.not) {
        return .js_undefined;
    }

    // handle failure

    const expected_args_js_array = try JSValue.createEmptyArray(globalThis, arguments.len);
    for (arguments, 0..) |arg, i| {
        try expected_args_js_array.putIndex(globalThis, @intCast(i), arg);
    }
    expected_args_js_array.ensureStillAlive();

    if (this.flags.not) {
        const signature = comptime getSignature("toHaveBeenLastCalledWith", "<green>...expected<r>", true);
        return this.throw(globalThis, signature, "\n\nExpected last call not to be with: <green>{any}<r>\nBut it was.", .{
            expected_args_js_array.toJestPrettyFormat(globalThis),
        });
    }
    const signature = comptime getSignature("toHaveBeenLastCalledWith", "<green>...expected<r>", false);

    if (totalCalls == 0) {
        return this.throw(globalThis, signature, "\n\nExpected: <green>{any}<r>\nBut it was not called.", .{
            expected_args_js_array.toJestPrettyFormat(globalThis),
        });
    }

    const diff_format = DiffFormatter{
        .expected = expected_args_js_array,
        .received = lastCallValue,
        .globalThis = globalThis,
        .not = false,
    };
    return this.throw(globalThis, signature, "\n\n{any}\n", .{diff_format});
}

const bun = @import("bun");
const DiffFormatter = @import("../diff_format.zig").DiffFormatter;

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/runtime/test_runner/expect/toHaveBeenLastCalledWith.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};

use super::mock;
use super::{Expect, get_signature};

pub(crate) fn to_have_been_last_called_with(
    this: &Expect,
    global: &JSGlobalObject,
    frame: &CallFrame,
) -> JsResult<JSValue> {
    bun_jsc::mark_binding!();
    let arguments = frame.arguments();
    let (this, calls, value) = this.mock_prologue(
        global,
        frame.this(),
        "toHaveBeenLastCalledWith",
        "<green>...expected<r>",
        mock::MockKind::CallsWithSig,
    )?;

    let total_calls: u32 = calls.get_length(global)? as u32;
    let mut last_call_value: JSValue = JSValue::ZERO;

    let mut pass = total_calls > 0;

    if pass {
        last_call_value = calls.get_index(global, total_calls - 1)?;

        if !last_call_value.js_type().is_array() {
            let mut formatter = super::make_formatter(global);
            return Err(global.throw(format_args!(
                "Expected value must be a mock function with calls: {}",
                value.to_fmt(&mut formatter),
            )));
        }

        pass = mock::call_args_equal(global, last_call_value, arguments)?;
    }

    if pass != this.flags.get().not() {
        return Ok(JSValue::UNDEFINED);
    }

    // handle failure
    let expected_args_js_array = JSValue::create_array_from_slice(global, arguments)?;
    expected_args_js_array.ensure_still_alive();

    if this.flags.get().not() {
        return mock::throw_not_failure(
            &this, global, "toHaveBeenLastCalledWith", "<green>...expected<r>",
            format_args!("Expected last call not to be with"), expected_args_js_array, "\nBut it was.",
        );
    }
    let signature = get_signature("toHaveBeenLastCalledWith", "<green>...expected<r>", false);

    if total_calls == 0 {
        return mock::throw_not_called(&this, global, signature, expected_args_js_array);
    }

    mock::throw_diff(&this, global, signature, format_args!(""), expected_args_js_array, last_call_value)
}
